File: 2020/day/10/src/Main.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE TupleSections #-}

module Main where

import Data.Foldable (traverse_)
import qualified Data.IntMap as Map
import Data.List (foldl', sort)
import System.Environment (getArgs)

main :: IO ()
main = (getArgs >>=) . traverse_ $ \filename -> do
  putStrLn filename

  adaptors <- map read . lines <$> readFile filename

  putStrLn "\tpart 1: product of 1-joltage differences and 3-joltage differences"
  putStrLn $ "\t" ++ show (part1 adaptors)

  putStrLn "\tpart 2: number of different adaptor chains"
  putStrLn $ "\t" ++ show (part2 adaptors)

type Joltage = Int

type JoltageMap = Map.IntMap

part1 :: [Joltage] -> Integer
part1 rs = (m Map.! 1) * (m Map.! 3)
  where
    m = differenceCounts rs

differenceCounts :: [Joltage] -> JoltageMap Integer
differenceCounts = Map.fromListWith (+) . map (,1) . (3 :) . (zipWith (-) =<< tail) . (0 :) . sort

-- dynamic programming!
part2 :: [Joltage] -> Integer
part2 = third . foldl' step (0, 0, 1, 0) . sort
  where
    third (_, _, c, _) = c
    step (!a, !b, !c, n) x = case x - n of
      1 -> (b, c, a + b + c, x)
      2 -> (c, 0, b + c, x)
      3 -> (0, 0, c, x)
      _ -> (0, 0, 0, x)

File: 2020/day/10/src/main.rs
use std::error;

type Error = Box<dyn error::Error>;

/*
macro_rules! error {
    ($fmt:literal $(, $arg:expr)*) => { Error::from(format!($fmt $(, $arg)* )) };
}
*/

fn main() -> Result<(), Error> {
    for filename in std::env::args().skip(1) {
        print!("{}", filename);

        let mut adaptors: Vec<Joltage> = std::fs::read_to_string(filename)?
            .lines()
            .map(|line| line.parse().map_err(Error::from))
            .collect::<Result<_, Error>>()?;
        adaptors.sort();

        println!("\tpart 1: product of 1-joltage differences and 3-joltage differences");
        println!("\t{}", part1(&adaptors));

        println!("\tpart 2: number of different adaptor chains");
        println!("\t{}", part2(&adaptors));
    }
    Ok(())
}

type Joltage = usize;

fn part1(adaptors: &[Joltage]) -> u64 {
    let mut counts = [0; 4];
    let mut prev = 0;

    for &r in adaptors {
        counts[r - prev] += 1;
        prev = r;
    }
    counts[3] += 1;

    return counts[1] * counts[3];
}

fn part2(adaptors: &[Joltage]) -> u64 {
    let mut a = 0; // ways to add to n - 2
    let mut b = 0; // ways to add to n - 1
    let mut c = 1; // ways to add to n
    let mut n = 0;

    for &r in adaptors {
        match r - n {
            1 => {
                let d = a + b + c;
                a = b;
                b = c;
                c = d;
            }
            2 => {
                let d = 0;
                let e = b + c + d;
                a = c;
                b = d;
                c = e;
            }
            3 => {
                let d = 0;
                let e = 0;
                let f = c;
                a = d;
                b = e;
                c = f;
            }
            _ => {
                a = 0;
                b = 0;
                c = 0;
            }
        }
        n = r;
    }

    c
}
